### PR TITLE
Castle Algorithm

### LINKED ISSUE #3140: [ENH] CASTLE Algorithm
**Is your feature request related to a problem? Please describe.**
CASTLE is a regularization method that improves supervised learning by jointly learning a DAG and a predictive model over the data

**Describe the solution you'd like A clear and concise description of what you want to happen, ideally taking into consideration the existing toolbox design, classes and methods.**
### `_CASTLEModel(nn.Module)` (Internal)
PyTorch masked autoencoder for feature reconstruction, target prediction, and training.

- **`__init__(num_inputs, model_cfg: ModelConfig, reg_cfg: RegularizationConfig)`**: Initializes masked input layers, scalar output layers, grouped hyperparameters, and optimizer.
- **`forward(X)`**: Returns full reconstruction (`Out`) and target prediction (`out_0`).
- **`train(X_tensor)`**: Runs epochs with early stopping (batch updates, Lagrangian multiplier adjustments), logs to TensorBoard if enabled, and returns the thresholded adjacency matrix (`W_final`).
- **`get_W()`**: Computes adjacency matrix from current weights (L2 norm of column $j$ in sub-network $i$).

### Internal dataclasses
Grouped configuration objects used by `_CASTLEModel`.

- **`ModelConfig`**: `hidden_dim`, `batch_size`, `max_epochs`, `optimizer`, `seed`, `min_loss_improvement`, `early_stop_patience`, `tensorboard_log_dir`, `scaler`, `target_col`
- **`RegularizationConfig`**: `dag_weight`, `sparsity_weight`, `dag_penalty`, `edge_threshold`

### `CASTLE(_BaseCausalDiscovery)` (Public API)
Validates data, orchestrates training, and builds the `pgmpy.DAG`.

```python
CASTLE( 
    dag_weight=1.0,
    sparsity_weight=5.0,
    dag_penalty=1.0,
    optimizer=None,
    batch_size=32,
    hidden_dim=32,
    edge_threshold=0.3,
    target_col=None,
    max_epochs=200, 
    min_loss_improvement=1e-4,
    early_stop_patience=10,
    scaler=None,
    tensorboard_log_dir=None,
    seed=None,
)
```

Joint training objective being minimized:

$$\min_\Theta \frac{1}{N}\|Y - [f_\Theta(\tilde{X})]_{:,1}\|^2 + \lambda \underbrace{\left(L_N(f_\Theta) + (\text{tr}(e^{M \odot M}) - d - 1)^2 + \beta V_{\Theta_1}\right)}_{R_{\text{DAG}}}$$

- `dag_weight` (λ): Weight on the entire DAG regularization loss $R_{\text{DAG}}$.
- `sparsity_weight` (β): Group-lasso penalty on input-layer weights $V_{\Theta_1}$, promoting edge sparsity.
- `dag_penalty` (ρ): Initial penalty coefficient in the Augmented Lagrangian for the acyclicity constraint.
- `optimizer`: Any `torch.optim.Optimizer` instance. Defaults to `Adam(lr=1e-3)` if `None`. Note: `weight_decay` is not recommended when passing a custom optimizer as CASTLE has its own sparsity mechanism via `dag_weight` and `sparsity_weight`.
- `batch_size`: Number of samples per mini-batch.
- `hidden_dim` (h): Hidden layer width in each sub-network $f_k$.
- `edge_threshold`: Edges with weight below this value are zeroed in the final DAG.
- `target_col`: Target column name or index. Defaults to the first column when `None`.
- `max_epochs`: Maximum number of training epochs.
- `min_loss_improvement`: Minimum decrease in total loss per epoch to count as an improvement for early stopping.
- `early_stop_patience`: Number of consecutive epochs without improvement before stopping early.
- `scaler`: Optional scaler for input normalization (for example, `sklearn.preprocessing.StandardScaler`). It must implement `fit(X)` and `transform(X)`; `inverse_transform(X)` is optional for user-side de-scaling. If provided, it is fit on training data and stored as `scaler_`.
- `tensorboard_log_dir`: If set, logs training metrics via `SummaryWriter(log_dir=...)`. If `None`, TensorBoard logging is disabled.
- `seed`: Seed for reproducibility.

**Methods:**
- **`fit(X)`**: Trains the model and builds the DAG.

**Attributes set after `fit`:**
- `n_features_in_`: Number of input features seen during fit.
- `feature_names_in_`: Feature names from the input DataFrame.
- `causal_graph_`: Learned causal DAG as a `pgmpy.base.DAG`.
- `adjacency_matrix_`: Weighted adjacency matrix as a Pandas DataFrame.
- `model_`: Internal neural network instance used for training.
- `scaler_`: Fitted scaler used to normalize inputs during training (when provided).

---
## Usage

```python
import torch
from pgmpy.causal_discovery import CASTLE
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

df = pd.DataFrame(np.random.randn(100, 3), columns=['A', 'B', 'Target'])

# Default: Adam with lr=1e-3
model = CASTLE(max_epochs=200, edge_threshold=0.3, seed=42)

# Custom optimizer — user controls all optimizer params directly
opt = torch.optim.Adam(lr=1e-4, weight_decay=1e-5)
model = CASTLE(max_epochs=200, edge_threshold=0.3, optimizer=opt, seed=42)

# User-provided scaler and early stopping defaults
scaler = StandardScaler()
model = CASTLE(
    max_epochs=200,
    edge_threshold=0.3,
    target_col="Target",
    scaler=scaler,
    min_loss_improvement=1e-4,
    early_stop_patience=10,
    tensorboard_log_dir="runs/castle",
    seed=42,
)
model.fit(df)
dag = model.causal_graph_

# User-side scaling and inverse-scaling for predictions (example)
scaler_fitted = model.scaler_
X_scaled = scaler_fitted.transform(df.values)
y_pred_scaled = some_model(X_scaled)
y_pred = scaler_fitted.inverse_transform(y_pred_scaled)
```

View logs with: `tensorboard --logdir runs/castle`

**Additional context**
paper: https://arxiv.org/pdf/2009.13180
ref codebase: https://github.com/trentkyono/CASTLE

Description: related : #2931, pgmpy/enhancement_proposals#4

Note this issue is part of an ongoing mentorship project

### PR DESCRIPTION
**Pull request by Manas-7854**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft. - Marked as draft
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md#ai-usage-policy)?
- [x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.  

- Please list the AI tool(s) used, along with the model and its version used.

GPT - 5.2 Codex

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

Provided the LLM with the exact API and step by step asked it to fill in the functions. At every step provided the LLM with a detailed understand of the task at hand and what it is supposed to implement in the function, used planning mode to review the planned implementation, and after the implementation was done review the code thoroughly.

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?  

Gone through the Code thoroughly, can tried to come up with test cases to cover almost every single functionality and tested that.

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

Gave the AI a list of tests to implement and check the code after the implementation. The number of tests has been reduced  bit.

### Issue number(s) that this pull request fixes
- Fixes #3140

### List of changes to the codebase in this pull request
- added `pgmpy/pgmpy/castle.py`
- updated `pgmpy/pgmpy/__init__.py` with the castle algorithm

### CODE REVIEW

**Comment by Manas-7854** [comment]
@ankurankan @DARHWOLF 
I have added this draft PR for `castle`, I will work on this and commit changes along the way. 
As of now to define the `optimizer` and `scalar` in the dataclass, I have used `object` - however this can be changed to `Any`, which will require importing `typing` library, however if we stick to `object`, we will have to call `isinstance` every single time we interact with it.
I would suggest switching to `Any`

File: pgmpy/tests/test_causal_discovery/test_castle.py [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+        raise ImportError("CASTLE requires PyTorch. Install it with: pip install torch")
+
+    monkeypatch.setattr(_castle, "_check_soft_dependencies", _raise)
+
+    with pytest.raises(ImportError, match="PyTorch"):
+        CASTLE()
+
+
+@requires_torch
+class TestCASTLEInit:
```
Comment (ankurankan) [inline pgmpy/tests/test_causal_discovery/test_castle.py]: Add the sklearn's `check_estimator` test.

File: pgmpy/tests/test_causal_discovery/test_castle.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+def numeric_df():
+    rng = np.random.default_rng(42)
+    return pd.DataFrame(rng.standard_normal((50, 3)), columns=["A", "B", "C"])
+
+
+class TestDataclasses:
+    def test_dataclasses_are_dataclasses(self):
+        assert dataclasses.is_dataclass(NetworkConfig)
+        assert dataclasses.is_dataclass(TrainingConfig)
+        assert dataclasses.is_dataclass(RegularizationConfig)
```
Comment (ankurankan) [inline pgmpy/tests/test_causal_discovery/test_castle.py]: Don't need to test this

File: pgmpy/causal_discovery/castle.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    Supervised causal discovery using CASTLE (CAusal STructure LEarning).
+
+    Jointly trains a neural network predictor and learns a causal DAG as an
+    auxiliary task via a masked autoencoder regularizer.
+
+    Parameters
+    ----------
+    dag_weight : float, default 1.0
+        Weight (λ) on the entire DAG regularization term (reconstruction loss +
+        acyclicity penalty + sparsity penalty).
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: Don't use unicode characters. Use the `math` directive, check numpydoc.

File: pgmpy/causal_discovery/castle.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        self.tensorboard_log_dir = tensorboard_log_dir
+        self.seed = seed
+
+    def _fit(self, X: pd.DataFrame):
+        """Fit the CASTLE model and construct the causal DAG."""
+        if X.shape[1] == 1:
+            raise ValueError("CASTLE requires at least 2 columns (one target and at least one feature).")
+        non_numeric = [col for col in X.columns if not pd.api.types.is_numeric_dtype(X[col])]
+        if non_numeric:
+            raise ValueError(f"All columns must be numeric. Non-numeric columns found: {non_numeric}.")
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: We have a `pgmpy.utils.preprocess_data` method for detecting variable types.

File: pgmpy/causal_discovery/castle.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+        if self.target_col is None:
+            target_col = X.columns[0]
+        elif isinstance(self.target_col, str):
+            if self.target_col not in X.columns:
+                raise ValueError(f"target_col '{self.target_col}' not found in DataFrame columns {list(X.columns)}.")
+            target_col = self.target_col
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py:309]: Why do we need to define a new variable here? Why not use `self.target_col` in the rest of the method?

File: pgmpy/causal_discovery/castle.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        else:
+            if not (0 <= self.target_col < X.shape[1]):
+                raise ValueError(
+                    f"target_col index {self.target_col} is out of range for DataFrame with {X.shape[1]} columns."
+                )
+            target_col = X.columns[self.target_col]
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: I don't think we should be supporting this.

File: pgmpy/causal_discovery/castle.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+        self.edge_threshold = edge_threshold
+        self.target_col = target_col
+        self.max_epochs = max_epochs
+        self.min_loss_improvement = min_loss_improvement
+        self.early_stop_patience = early_stop_patience
+        self.scaler = scaler
+        self.tensorboard_log_dir = tensorboard_log_dir
+        self.seed = seed
+
+    def _fit(self, X: pd.DataFrame):
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py:296]: Try to add comments with `Step x`. Have a look at our other algorithms. Looking at the step comments should be able to give us the pseudo-code of the algorithm.

File: pgmpy/causal_discovery/castle.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+            tensorboard_log_dir=self.tensorboard_log_dir,
+        )
+        self.reg_config_ = RegularizationConfig(
+            dag_weight=self.dag_weight,
+            sparsity_weight=self.sparsity_weight,
+            dag_penalty=self.dag_penalty,
+            edge_threshold=self.edge_threshold,
+        )
+
+        self.cols_ = [target_col] + [c for c in X.columns if c != target_col]
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: Is this different from `self._features_names_in_`?

File: pgmpy/causal_discovery/castle.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+        self.predictor_names_ = self.cols_[1:]
+        X = X[self.cols_]
+
+        self.scaler_ = self.scaler if self.scaler is not None else StandardScaler()
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: This should go to the top of this method. This is a preprocessing step; pre-processing should be done in `Step 0` before any actual algorithm steps start.

File: pgmpy/causal_discovery/castle.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+        X_scaled = self.scaler_.fit_transform(X.to_numpy())
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: I don't think we need a `to_numpy` call. sklearn works with dataframe objects.

File: pgmpy/causal_discovery/castle.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+
+        X_tensor = torch.tensor(X_scaled, dtype=torch.float32)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: dtype and device need to be read in from `pgmpy.global_config`.

File: pgmpy/causal_discovery/castle.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+
+        self.model_ = _CASTLEModel(len(self.cols_), self.network_config_, self.train_config_, self.reg_config_)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: `len(self.cols_)` should be equivalent to `self.n_features_in_` which is already set in the base class. We should use that as that is the standard way for sklearn estimators.

File: pgmpy/causal_discovery/castle.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+        W_final = self.model_.train(X_tensor)
+
+        self.adjacency_matrix_ = pd.DataFrame(W_final.cpu().numpy(), index=self.cols_, columns=self.cols_)
+
+        self.causal_graph_ = DAG()
+        self.causal_graph_.add_nodes_from(self.cols_)
+        for src in self.cols_:
+            for dst in self.cols_:
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: Using `itertools.combination` would be cleaner than doing nested loops here.

File: pgmpy/causal_discovery/castle.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+                if self.adjacency_matrix_.loc[src, dst] > 0:
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: Shouldn't this be using the `edge_threshold` argument?

File: pgmpy/causal_discovery/castle.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+    "sgd": {"lr", "weight_decay", "momentum", "dampening", "nesterov", "maximize"},
+    "adamw": {"lr", "betas", "weight_decay", "amsgrad", "maximize", "eps"},
+}
+
+
+def _validate_optimizer(optimizer: str, optimizer_kwargs: dict) -> None:
+    """Validate optimizer name and kwargs, raising ValueError on any invalid input."""
+    name = optimizer.lower()
+    if name not in _OPTIMIZER_VALID_KWARGS:
+        valid = ", ".join(f"'{k}'" for k in sorted(_OPTIMIZER_VALID_KWARGS))
```
Comment (ankurankan) [inline pgmpy/causal_discovery/castle.py]: I don't see why we need to construct a string here. Just passing `_OPTIMIZER_VALID_KWARGS.keys()` in the next print statement should give the same output?

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/causal_discovery/castle.py (hunk 13 above)
Comment (Manas-7854) [reply pgmpy/causal_discovery/castle.py]: Using `itertools.combination` would not work since we need to preserve the ordering and `itertools.combination` does not do that.
We can use `itertools.permutation` instead - I will make the change in the code

File: pgmpy/causal_discovery/castle.py (hunk 14 above)
Comment (Manas-7854) [reply pgmpy/causal_discovery/castle.py]: The constraint is already applied inside `_CASTLE_MODEL.train` (around line 183)

File: pgmpy/causal_discovery/castle.py (hunk 5 above)
Comment (Manas-7854) [reply pgmpy/causal_discovery/castle.py:309]: Yes, we can just use `self.target_col` instead of defining a new variable but that would require adding a similar check every time we use the variable. If it is `None` - `target_col` should be `X.columns[0]` and `self.target_cols` otherwise.
Defining a new variable would be a cleaner way to do this.

**Review by ankurankan — CHANGES_REQUESTED** [review]
Could you compare this implementation with the official implmenetation?
